### PR TITLE
Pulls tweets

### PR DESCRIPTION
**Pull request by davidfurlong**
TODO's

- [x] On signup, should populate a new graphdoc with the users most recent 200 tweets (tweets as nodes)

- [x] This graphdoc should update continuously with new tweets each time you log in.

- An 'import tweets' button in every cograph that lets you import some number n of your tweets to whatever cograph you are looking at. This import should also be able to be filtered. I.e. hashtag + which groups of people wrote them): MOVED TO: https://github.com/willzeng/cograph/pull/630

### CODE REVIEW

File: config/passport.js [hunk 1]
```diff
@@ -36,9 +38,14 @@ module.exports = function(passport) {
 
     // Twitter Signup
     passport.use(new TwitterStrategy({
+        // remote
         consumerKey: "iRnrLu6QrYHPlOF0wq2ns1MYl",
         consumerSecret: "bdIQkb16hSVAvr64sTkq0YXhyysBoZ5dvMQSM9d3tdsCz3JdNx",
         callbackURL: "http://www.cograph.co/auth/twitter/callback"
+        // local
```
Comment (willzeng) [inline config/passport.js]: @davidfurlong What is the difference between these two sets of credentials?

Comment (davidfurlong) [inline config/passport.js]: Callbacks from apps.twitter.com config have to match

File: config/passport.js [hunk 2]
```diff
@@ -61,22 +68,43 @@ module.exports = function(passport) {
...
+                            var newUser             = new User();
+                            // set the user's local credentials
+                            newUser.local.email     = profile.email;
+                            newUser.local.name      = profile.username;
+                            newUser.local.nameLower = profile.username.toLowerCase();
+                            newUser.twitter = profile._json;
+                            newUser.twitter.id = profile.id;
+                            newUser.twitter.username = profile.username;
+                            newUser.twitter.displayName = profile.displayName;
+                            newUser.twitter.tweets = data; // 200 tweets
```
Comment (willzeng) [inline config/passport.js]: @davidfurlong Am I reading it correctly that this is where the users last 200 tweets are stored? 

Just to check that I'm getting this right.  This is a one time pull of 200 tweets at startup that then stores the tweet data in our mongoDb and does not update it?

Comment (davidfurlong) [inline config/passport.js]: Yep

**Comment by willzeng** [comment]
Let's write up to exactly specify what we want this feature to do. Some suggestions.
- On signup, should populate a new graphdoc with the users most recent 200 tweets (tweets as nodes)
- This graphdoc should update continuously with new tweets after we have been added to their twitter.  @davidfurlong Do you know how to set this up second part up with the twitter api?

File: config/passport.js (hunk 1 above)
Comment (willzeng) [inline config/passport.js]: So one is for testing locally and one is for deploymment? I'm looking at the callback URLs

Comment (davidfurlong) [inline config/passport.js]: Yep

**Comment by davidfurlong** [comment]
@willzeng this should work now

**Comment by willzeng** [comment]
Yep that works for me.  Now we need to take that string and turn it into a cograph that updates each time the use logs in?

**Comment by willzeng** [comment]
@davidfurlong Rad.  So I've got the first check box ticked off for this PR.  When you sign up with twitter for the first time it takes your most recent 200 tweets and drops them into an automatically created cograph.  The next step could have two paths:

We could make it so that everytime you log in with twitter, your recent tweets are merged into that special "tweet" cograph.  
And/OR
We could add an 'import tweets' button to every cograph that lets your grab your most recent 'n' tweets and adds them to that cograph.

What do you think?

**Comment by davidfurlong** [comment]
Flexibility is probably best. One graph that has your tweets / people you
follows? Then ability to add twitter "landscapes" based on a filter as an
import button on a graph

**Comment by willzeng** [comment]
Ok so basically both.  I think the functionality is quite similar so doing one and then the other should be relatively straightforward.

**Comment by willzeng** [comment]
@davidfurlong See todo's at top

**Comment by willzeng** [comment]
Got the update on login functionality working @davidfurlong Check it out.

**Comment by willzeng** [comment]
Maybe the thing to do for import is to just let people type in twitter search queries and we just give them whatever results come in?

**Comment by davidfurlong** [comment]
> Maybe the thing to do for import is to just let people type in twitter search queries and we just give them whatever results come in?

yeah thats cool + easy to do

**Comment by willzeng** [comment]
Lets plan to get that and some automatic linking of them done before
pushing twitter integration live on Wednesday.

**Comment by willzeng** [comment]
Yep, I don't have this bug.  We'll try and replicate when you get here.

**Comment by davidfurlong** [comment]
the last todo is on branch nsearch, just needs functionality

**Comment by willzeng** [comment]
@davidfurlong The update is now fixed so that there is an auto updating functional cograph of a users tweets.  Give this a check to make sure it is also working with your account.

If so, we can change the number so that initial signup pulls 200 tweets and each login checks the most recent 200? (it automatically avoids making duplicates)  This should then be merge-able.

I created a new pull request for the query adder box: https://github.com/willzeng/cograph/pull/630

**Comment by willzeng** [comment]
- [x] Just realized one additional thing to do is to check (on update) to make sure there is a twittercograph so that the upgrade works

**Comment by willzeng** [comment]
This is now working and stable, though when looking at the larger tweet cographs it makes me think that we should really be doing some optimizations.

**Comment by davidfurlong** [comment]
can you post an example? what kind of optimizations

**Comment by davidfurlong** [comment]
are we still storing it in the mongo (probably not a good idea)

**Comment by willzeng** [comment]
Ah yes we are, and needlessly so. Go ahead and remove that.

I mean that graphview gets a little slow with 200 nodes and lots of images.
Set TweetNumber to 100 and give it a try.

We should probably do some processing of the images to be smaller res when
in the graph. This issue is seperate from this PR.

**Comment by davidfurlong** [comment]
Are we good to go?

**Comment by willzeng** [comment]
If it worked for your account then yeah.  We can then do another update
that includes the functional custom search box import feature.

**Pull request merged**